Abort after failing C assert

tautschnig opened the pull request:
The C standard specifies behaviour of the "assert" macro as resulting in
an abort when the condition does not evaluate to true. Implement this
behaviour by inserting assume(0) after assert(0).

Fixes: https://github.com/diffblue/cbmc/issues/5505

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

peterschrammel commented on regression/cbmc-incr-oneloop/multiple-asserts/test.c:
```diff
@@ -2,8 +2,8 @@ int main()
   {
-    assert(i != 5);
-    assert(i != 8);
+    __CPROVER_assert(i != 5, "");
```
❓ Do we need a convenienve overload without the comment argument since `__CPROVER_assert` is now the default way to specify non-aborting properties?

hannes-steffenhagen-diffblue replied:
Sounds like a good idea to me

peterschrammel reviewed (commented):
I think this requires a thorough documentation review since `__CPROVER_assert` is now the default way to specify non-aborting properties. Some downstream CBMC users might not operate with abort-semantics in mind.

Side note: JBMC does something similar (in one of its modes) https://github.com/diffblue/cbmc/blob/f4edb72f3db37f039c215d35bb0d75bd80540742/jbmc/src/java_bytecode/java_bytecode_convert_method.cpp#L2444-L2446

hannes-steffenhagen-diffblue commented:
IMHO this sounds like the right way to go forward with this. I do also agree with @peterschrammel though that this requires us going through documentation and making sure we don't say anything wrong now, and for the other I think this should result in a major version bump.

TGWDB commented on regression/goto-harness/do-not-use-nondet-for-selecting-pointers-to-treat-as-equal/test.desc:
```diff
@@ -2,11 +2,11 @@ CORE
-\[test.assertion.3\] line 5 assertion x == y: FAILURE
-\[test.assertion.4\] line 6 assertion x != y: FAILURE
-\[test.assertion.5\] line 7 assertion \*x == \*y: FAILURE
+\[test.assertion.1\] line 3 assertion: SUCCESS
+\[test.assertion.2\] line 4 assertion: SUCCESS
+\[test.assertion.3\] line 5 assertion: FAILURE
+\[test.assertion.4\] line 6 assertion: FAILURE
+\[test.assertion.5\] line 7 assertion: FAILURE
```
Have we lost something useful and helpful here by not showing any detail of the assertion that failed? (I realise the empty messages may contribute to this, but with the comment of @peterschrammel above perhaps we should consider having the same behaviour in our override?)

martin-cs replied:
+1 these should give a message...

martin-cs commented:
We are doing a similar assert/assume thing for Ada.

As regards updating -- this always used to be how `assert()` worked.  I guess it must have been removed at some point?  There were a few years where there was ... ambiguity over the semantics of assert.

martin-cs reviewed (approved):
A good and useful thing to do.  I won't block this but I do agree with others that this needs to be clearly communicated (or, at least, "If you want a non-terminating check, use `__CPROVER_assert()`").

One thought : it feels like `assert` should be implementable in the C library (possibly with a macro to copy the expression into the string so that we don't have the problem with loss of messages that @TGWDB flagged).  I am generally in favour of modelling going in the library in C where possible.

SaswatPadhi commented:
I have one small comment that I had mentioned to @tautschnig this morning:

The semantics for `ASSUME` statement currently is stated (in [`src/goto-programs/goto_program.h`](https://github.com/diffblue/cbmc/blob/develop/src/goto-programs/goto_program.h)) as:
> This thread of execution waits for `guard` to evaluate to true.
> Assume does not "retro-actively" affect the thread or any ASSERTs.

This means, if the `guard` evaluates to `false`, it would keep waiting forever as opposed to aborting (as mandated by C `assert` semantics).

It seems to me that `THROW` might be the closest we have to an aborting goto primitive, but I'm not sure.

martin-cs commented:
@SaswatPadhi This is a sensible and a valid concern but I think that `assume` is the right thing to do.  Our (informal) model of execution has nothing happen after termination, so in the single threaded case, an abrupt termination and an infinite loop are indistinguishable.  The multi-threaded case is a little more complex and I haven't worked on it so I am probably not the best person to ask but I don't think we have a way of distinguishing a "hung" thread from one that terminates early, so, again, this is not an issue.

`THROW`and `CATCH`... TBH their semantics are even less clear to me.  I think the clearest explanation is the code that removes them.  This is different for C++ and Java.  In practice they are removed early on so I don't think we should be introducing them, esp. in things like C that don't have an exception model at all.

tautschnig committed:
Regression tests: use __CPROVER_assert when multiple properties may fail

C's assert() should result in an abort() when the asserted condition
evaluates to false. If we want to test multiple properties, including
some failing ones, use __CPROVER_assert to not rely on any modelling of
assert() that does not result in an abort().

tautschnig committed:
Abort after failing C assert

The C standard specifies behaviour of the "assert" macro as resulting in
an abort when the condition does not evaluate to true. Implement this
behaviour by inserting assume(0) after assert(0).

Fixes: #5505

tautschnig commented:
Marking do-not-merge as this likely wants a major-version bump. Otherwise it should be ready for (re-)review.

feliperodri commented:
Fixes: https://github.com/diffblue/cbmc/issues/5505

esteffin commented:
We are considering to include this PR into version 6.

However we should make sure that we do add the documentation specifying that `assert` is blocking as opposed to `__CPROVER_assert` that is not.

tautschnig commented:
This is a major undertaking with limited added value. Users could choose to add their own assert macro that exhibits this behaviour. Closing as we are unlikely to find time to work on this anytime soon.
